weave/trace_server/execution_runner/cross_process_trace_server.py

```diff
@@ -0,0 +1,540 @@
+
+    # Start the worker thread IN THE PARENT PROCESS
+    worker_thread = threading.Thread(
+        target=_trace_server_worker_loop_with_context,
+        args=(request_queue, response_queue, trace_server, context_values),
+        daemon=True,
+    )
+    worker_thread.start()
+
+    # Return ONLY the queues (not the trace server!)
+    return CrossProcessTraceServerArgs(
+        request_queue=request_queue, response_queue=response_queue
+    )
+
+
+def build_child_process_trace_server(
+    args: CrossProcessTraceServerArgs,
+) -> CrossProcessTraceServer:
+    """
+    Build a CrossProcessTraceServer from serializable arguments.
+
+    This function runs in the CHILD process with queues created in the parent.
+
+    Args:
+        args: The arguments containing the communication queues from parent
+
+    Returns:
+        A CrossProcessTraceServer instance that communicates with the parent's worker
+    """
+    return CrossProcessTraceServer(
+        request_queue=args["request_queue"], response_queue=args["response_queue"]
+    )
+
+
+def _trace_server_worker_loop_with_context(
+    request_queue: multiprocessing.Queue,
+    response_queue: multiprocessing.Queue,
+    trace_server: TraceServerInterface,
+    context_values: dict[contextvars.ContextVar, Any],
+) -> None:
+    """
+    Worker loop that handles requests using the provided trace server with context.
+
+    This function runs in a separate thread started by generate_child_process_trace_server_args.
+
+    Args:
+        request_queue: Queue to receive requests from
+        response_queue: Queue to send responses to
+        trace_server: The actual trace server instance to delegate to
+        context_values: The current context variable values to propagate
+    """
+    # Set context variables in the worker thread
+    for var, value in context_values.items():
+        if value is not None:  # Only set if there was a value
+            var.set(value)
+
+    # Now run the normal worker loop
+    while True:
+        try:
+            # Get the next request
+            wrapped_request = request_queue.get()
+
+            # Handle shutdown signal
+            if wrapped_request["method"] == "_shutdown":
+                break
+
+            # Process the request
+            try:
+                method = getattr(trace_server, wrapped_request["method"])
+                result = method(wrapped_request["request"])
+
+                response = ResponseWrapper(
+                    request_id=wrapped_request["request_id"], result=result, error=None
+                )
+            except Exception as e:
+                response = ResponseWrapper(
+                    request_id=wrapped_request["request_id"], result=None, error=str(e)
+                )
+
+            # Send the response
+            response_queue.put(response)
+
+        except Exception as e:
+            # Critical error in worker loop
+            logger.exception(f"Critical error in worker loop: {e}")
```

Comment:
nit: the double try/except is a bit gnarly, suggest making a helper fn

Reply:
this is not a double try/except. maybe I am missing something?

Reply:
this should be fixed now